---
title: "Maryland Genomics Illumina Report"
output: 
  pdf_document:
    latex_engine: xelatex
    fig_caption: FALSE
    extra_dependencies: ["float"]
mainfont: DejaVu Sans Mono
geometry: margin=0.5cm
params:
  projectname: "Project"
  thedate: !r Sys.Date()
  pathd: ""
  projid: "A01_1"
  runid: "20220313_S64049_1"
  basep: "/local/sequence/illumina"  
header-includes:
   -  \usepackage{colortbl} 
   -  \usepackage{color} 
   -  \usepackage{xcolor}
   -  \usepackage{caption}
   -  \captionsetup[table]{labelformat=empty}
   -  \usepackage{float}
   -  \usepackage{fontspec}
   -  \usepackage{dejavu}
   -  \setmainfont{DejaVu Sans Mono}
---

```{r setup, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE)
knitr::opts_chunk$set(fig.pos = 'H')
knitr::opts_chunk$set(tab.cap.pre = "", tab.cap.sep = "")
knitr::opts_chunk$set(echo = FALSE, ft.keepnext = F)
knitr::opts_chunk$set(tab.cap.pre = "", tab.cap.sep = "")
```

```{r setup2, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
require(XML)
#classoption: landscape
mytheme <- gridExtra::ttheme_default(base_size=18, core = list(fg_params=list(cex = 0.75, fontface="plain")),colhead = list(fg_params=list(cex = 0.75, fontface="bold")),rowhead = list(fg_params=list(cex = 0.75, fontface="bold")))
mytheme1 <- gridExtra::ttheme_default(base_size=18, core = list(fg_params=list(cex = 0.75, fontface="plain")),colhead = list(fg_params=list(cex = 0.75, fontface="bold")),rowhead = list(fg_params=list(cex = 0.75, fontface="bold")))
mytheme2 <- gridExtra::ttheme_default(base_size=15, core = list(fg_params=list(cex = 0.75, fontface="plain")),colhead = list(fg_params=list(cex = 0.75, fontface="bold")),rowhead = list(fg_params=list(cex = 0.75, fontface="bold")))
library(gridExtra)
library(dplyr)
library(grid)
library(cowplot)
library(magick)
library(ggplot2)
library(flextable)
library(webshot)
odirec=unlist(params$pathd)

#param for local tests
#runid="220826_A00904_0245_BHFGFKDMXY"
#projid="COVID"
#Look at local copy for test paths.

runid=unlist(params$runid)
projid=unlist(params$projid)
basep=unlist(params$basep)

#Path to instruments file.
inst_p="/local/projects/grc/pipelines_conf/instruments.conf"


runpath=paste0(basep,"/",runid,"/")
r_param=paste0(runpath, "/RunParameters.xml")
qcplots=paste0(runpath,"/PA_Logs/")

#create directory to write tables to
csvout=paste0(odirec,"/report_tables")
if (!dir.exists(csvout)){
    dir.create(csvout)
}
#set_table_properties(layout = "autofit")
```

```{r logo, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}
knitr::include_graphics("/local/projects-t3/achatterjee/MDG/md-genomics-horizontal.pdf")

```
This report contains run information and quality metrics for an Illumina run generated by Maryland Genomics. If the project requires
multiple flowcell runs, a separate report will be generated for each flowcell. For additional information or questions, please
contact IGS-Services@SOM.UMaryland.edu

```{r sampledat, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", results='asis', fig.pos='H'}
set_flextable_defaults(font.family = 'DejaVuSansMono')

runinf=paste0(qcplots,"/runinfo.csv")
sampletab=read.csv(runinf, skip=1, header = T, row.names = NULL)
cnames=colnames(sampletab)[2:length(colnames(sampletab))]
sampletab$index_sequence=NULL
colnames(sampletab)=cnames
proj_samples=sampletab[sampletab$project==projid,]
lanes=unlist(unique(proj_samples$lane))
      if(length(lanes)>1){
        lane=paste0(lanes, collapse = ",")
      } else {
        lane=toString(lanes)
      }
proj_samples$sequencing_sub_sample_id<-NULL
proj_samples$loaded<-NULL
proj_samples$needed<-NULL
proj_samples$mp_insert_size<-NULL
proj_samples$median_smear_size<-NULL

# Reorg columns and change colnames
#1,9,2, 3,10, 5, 7,8,6,11,12,4,13

cnames=c("Lane","Project","Master Sample ID", "Library ID", "Sample Name","External ID","Type","Species","Strain","Method","Index Type","Index Number","Index Sequence")

proj_samplesf=proj_samples[,c(1,9,2,3,10,5,7,8,6,11,12,4,13)]
colnames(proj_samplesf)<-cnames
if(all(is.na(proj_samplesf$Strain))){
  proj_samplesf$Strain=NULL
}
proj_samplesff<-proj_samplesf[rowSums(is.na(proj_samplesf)) != ncol(proj_samplesf),]
proj_samplesf<-proj_samplesff

#flextable
ft_sampledat=autofit(flextable(proj_samplesf)%>%set_caption(caption = paste0("Table 1: Sample summary table for samples from project ",projid), autonum = FALSE)%>% theme_zebra() %>% bg(bg = "#44B29C", part = "header")) %>%fit_to_width(max_width = 8.5)

#ft_sampledat %>% fit_to_width(8)
flextable_to_rmd(ft_sampledat, ft.split = TRUE)

write.csv(proj_samplesf, file=paste0(csvout, "/table1.csv"))
```

```{r rinfo, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", results='asis'}
rif=paste0(runpath,"/RunInfo.xml")
data <- xmlParse(rif)
xml_data <- xmlToList(data)
recipe_xml=xml_data$Run$Reads
recipe_t=""
for(r in recipe_xml){
  recipe_t=paste0(recipe_t,":", r[2])
}
recipe=substring(recipe_t, 2)

###Specs from Ivette

instrument_code=unlist(strsplit(runid, "_"))[2]
inst_look=read.csv(inst_p, comment.char = "#", header=F)
hold=inst_look[inst_look$V2==instrument_code,]
Inst_type=hold$V3

dat <- xmlParse(r_param)
xml_data <- xmlToList(dat)
flowcell_type=xml_data$RfidsInfo$FlowCellMode
if(is.null(flowcell_type)==TRUE){
  flowcell_type="Illumina MiSeq"
}
hdat=data.frame("InstrumentType"=Inst_type, "FlowCellType"=flowcell_type, "RunRecipe"=recipe)
#flextable convert
gt1=autofit(flextable(hdat)%>%set_caption(caption = "Table 2:  Run info, including instrument type, run & flowcell type", autonum = FALSE)%>% theme_zebra() %>% bg(bg = "#44B29C", part = "header")) %>%fit_to_width(max_width = 8.5)

#ft_sampledat %>% fit_to_width(8)
flextable_to_rmd(gt1)
write.csv(hdat, file=paste0(csvout, "/table2.csv"))
```

```{r Figures1a, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="Flowcell Passed-Filter Cluster Count Distribution by Lane."}
newplots=paste0(qcplots,"/images/")
laneplots=list()
if(length(lanes)==1){
fig1n=unlist(list.files(path=newplots, pattern=paste0(lanes,"_ClusterCount-by-lane.png"), full.names = T))
fig1 = fig1n[file.size(fig1n) != 0L]
if(file.exists(fig1)){

ggdraw() + draw_image(fig1)
}}

if(length(lanes)>1){
  for(l in lanes){
    fig1n=unlist(list.files(path=newplots, pattern=paste0(l,"_ClusterCount-by-lane.png"), full.names = T))
    fig1 = fig1n[file.size(fig1n) != 0L]
    if(file.exists(fig1)){
      laneplots=rbind(laneplots,  fig1)
    }
  }
  laneplots=unlist(laneplots)
  lp=image_read(laneplots)
  image_montage(lp, tile=round(length(lp)/2),geometry = 'x400+10+10', gravity = "Center")
  }

```

```{r Figures1b, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="Flowcell Signal Intensity By Lane"}
laneplots=list()
if(length(lanes)==1){
fig2n=unlist(list.files(path=newplots, pattern=paste0(lanes,"_flowcell-Intensity.png"), full.names = T))
fig2 = fig2n[file.size(fig2n) != 0L]
if(file.exists(fig2)){
#if(file.size(fig1) != 0L){
ggdraw() + draw_image(fig2)
}}

if(length(lanes)>1){
  for(l in lanes){
    fig2n=unlist(list.files(path=newplots, pattern=paste0(l,"_flowcell-Intensity.png"), full.names = T))
    fig2 = fig2n[file.size(fig2n) != 0L]
    if(file.exists(fig2)){
      laneplots=rbind(laneplots,  fig2)
    }
  }
  laneplots=unlist(laneplots)
  lp=image_read(laneplots)
  image_montage(lp, tile=round(length(lp)/2),geometry = 'x400+10+10', gravity = "Center")
  }
```

```{r Figures1c, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="Flowcell Signal Intensity By Cycle"}
laneplots=list()
if(length(lanes)==1){
fig3n=unlist(list.files(path=newplots, pattern=paste0(lanes,"_Intensity-by-cycle_Intensity.png"), full.names = T))
fig3 = fig3n[file.size(fig3n) != 0L]
if(file.exists(fig3)){
#if(file.size(fig1) != 0L){
ggdraw() + draw_image(fig3)
}}

if(length(lanes)>1){
  for(l in lanes){
    fig3n=unlist(list.files(path=newplots, pattern=paste0(l,"_Intensity-by-cycle_Intensity.png"), full.names = T))
    fig3 = fig3n[file.size(fig3n) != 0L]
    if(file.exists(fig3)){
      laneplots=rbind(laneplots,  fig3)
    }
  }
  laneplots=unlist(laneplots)
  lp=image_read(laneplots)
  image_montage(lp, tile=round(length(lp)/2),geometry = 'x400+10+10', gravity = "Center")
  }
```

```{r Figures1d, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="Flowcell Mean Quality Score Distribution By Cycle"}
laneplots=list()
if(length(lanes)==1){
fig4n=unlist(list.files(path=newplots, pattern=paste0(lanes,"_q-heat-map.png"), full.names = T))
fig4 = fig4n[file.size(fig4n) != 0L]
if(file.exists(fig4)){
#if(file.size(fig1) != 0L){
ggdraw() + draw_image(fig4)
}}

if(length(lanes)>1){
  for(l in lanes){
    fig4n=unlist(list.files(path=newplots, pattern=paste0(l,"_q-heat-map.png"), full.names = T))
    fig4 = fig4n[file.size(fig4n) != 0L]
    if(file.exists(fig4)){
      laneplots=rbind(laneplots,  fig4)
    }
  }
  laneplots=unlist(laneplots)
  lp=image_read(laneplots)
  image_montage(lp, tile=round(length(lp)/2),geometry = 'x400+10+10', gravity = "Center")
  }
```


```{r qcmetrics, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", results='asis', out.width='100%'}
summary=paste0(qcplots,"/summary.csv")
sumtext=readLines(summary)
i1=which(!nzchar(sumtext)) [1] 
i1=i1-1
readsum=sumtext[3:i1]
#gsub("\"","",readsum)
summary_table=read.table(text=readsum,header=T, fill = T, as.is = T, sep=",", check.names = F)

#flextable convert
gt1=autofit(flextable(summary_table)%>%set_caption(caption = "Table 3: Flowcell QC Metrics", autonum = FALSE)%>% theme_zebra() %>% bg(bg = "#44B29C", part = "header")) %>%fit_to_width(max_width = 8)

#ft_sampledat %>% fit_to_width(8)
flextable_to_rmd(gt1)
write.csv(summary_table, file=paste0(csvout, "/table3.csv"))
```


```{r demux1, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", results='asis'}
##mapping file implementstion
demux_map<-paste0(qcplots, "demux_map_table.csv")
mapf<-read.table(demux_map, sep=",", header=T)
pmaps<-mapf[mapf$project==projid,]
t1_list<-data.frame()
t2_list<-data.frame()
t1f=FALSE
t2f=FALSE
t3f=FALSE
samples<-proj_samples$sample_id
smapf<-mapf[mapf$sample_id %in% samples,]

exists<-list()
for(path in smapf$demux_file){
  if(file.exists(path)){
    exists<-c(exists,path)
  }
}
existing_files<-unlist(exists)

for(path in existing_files){
  alltables=readHTMLTable(path)
  for(i in alltables){
  if(is.null(i)==F){
    if("Project" %in% colnames(i)){
      hold=i[rowSums(is.na(i)) != ncol(i),]
      proj=paste0("Project_",projid)
      tablef=hold[hold$Project==proj,]
      tablef$`PF Clusters`<-as.numeric(gsub(",","",tablef$`PF Clusters`))
      tablef$`Yield (Mbases)`<-as.numeric(gsub(",","",tablef$`Yield (Mbases)`))
      t1_list<-rbind(t1_list,tablef)
      yield=sum(tablef$`Yield (Mbases)`)
      lanes=unlist(unique(tablef$Lane))
      if(length(lanes)>1){
        lane=paste0(lanes, collapse = ",")
      } else {
        lane=toString(lanes)
      }
      #lane=paste(shQuote(lanes), collapse=", ")
      pfcluster=sum(tablef$`PF Clusters`)
      tab1=data.frame("Clusters(PF)"=prettyNum(pfcluster,big.mark=",",scientific=FALSE), "Yield (MBases)"=prettyNum(yield,big.mark=",",scientific=FALSE), check.names = F)
      #tab1<-prettyNum(tab11, big.mark=",",scientific=FALSE)
    }}}}
if(is.null(existing_files)==F){
tablef<-unique(t1_list)

pfcluster=sum(tablef$`PF Clusters`)
      tab1=data.frame("Clusters(PF)"=prettyNum(pfcluster,big.mark=",",scientific=FALSE), "Yield (MBases)"=prettyNum(yield,big.mark=",",scientific=FALSE), check.names = F)
      ##edit table content.
      tablef$Project<-gsub("Project_","", tablef$Project)
      tablef$`Barcode sequence`<-NULL
      #Add summary row at bottom
      #Add summary row with sums of PF Clusters, % of the lane, and Yield. And averages of the remaining columns
      tablef[c(4,5,6,7,8,9,10,11)]<-sapply(tablef[c(4,5,6,7,8,9,10,11)],as.numeric)
      table4<-tablef %>% bind_rows(summarise(.,
                        across(where(is.numeric), sum),
                        across(where(is.character), ~"Total")))
      nums=nrow(tablef)
      lastr=nums+1
      table4$`Mean QualityScore`[lastr]<-table4$`Mean QualityScore`[lastr]/nums
      table4$`% Perfectbarcode`[lastr]<-table4$`% Perfectbarcode`[lastr]/nums
      table4$`% One mismatchbarcode`[lastr]<-table4$`% One mismatchbarcode`[lastr]/nums
      table4$`% PFClusters`[lastr]<-table4$`% PFClusters`[lastr]/nums
      table4$`% >= Q30bases`[lastr]<-table4$`% >= Q30bases`[lastr]/nums
      t4_sampleID<-merge(table4, proj_samplesf[, c("Library ID", "Master Sample ID")], by.x="Sample", by.y = "Library ID", all.x=T)
      table4f<-t4_sampleID[,c(1,12,2,3,4,5,6,7,8,9,10,11)]
      names(table4f)[names(table4f) == 'Sample'] <- 'Library ID'
      names(table4f)[names(table4f) == 'Master Sample ID'] <- 'Sample ID'
      table4f$`Sample ID`[nrow(table4f)]<-"Total"
      #flextable
      t1f=autofit(flextable(tab1)%>%set_caption(caption = paste0("Table 4a: Total Passed-Filter Data Generated for Project ", projid), autonum = FALSE)%>% theme_zebra() %>% bg(bg = "#44B29C", part = "header")) %>%fit_to_width(max_width = 8.5)
      flextable_to_rmd(t1f)
      t2f=autofit(flextable(tablef)%>%set_caption(caption = "Table 4b: Per-Sample Summary Statistics", autonum = FALSE)%>% theme_zebra() %>% bg(bg = "#44B29C", part = "header")) %>%fit_to_width(max_width = 8.5)
      flextable_to_rmd(t2f)
      write.csv(tab1, file=paste0(csvout, "/table4a.csv"))
      write.csv(tablef, file=paste0(csvout, "/table4b.csv"))
}
```

```{r trimstats, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center",fig.width=20, fig.height=14, results='asis'}
#tstats="/Users/apaala.chatterjee/MDG/Illumina/trimming.logs.csv"
tstats=paste0(qcplots, "/trimming.logs.csv")
trimstats=read.csv(tstats, header = T, row.names = NULL)
trimf=trimstats[trimstats$Project==projid,]
trimf$Lane<-gsub("L00","",trimf$Lane)
trimtable<-trimf[,colSums(is.na(trimf))<nrow(trimf)]
####flextable
if(ncol(trimtable)==15){
ft_runinfo=autofit(flextable(trimtable)%>%set_caption(caption = "Table 5: Adaptor and Quality Trimming Summary")%>% theme_zebra()%>%bg(bg = "#44B29C", part = "header"))%>%add_header_row( values = c("","Read 1 Untrimmed","","Read 2 Untrimmed",""),colwidths = c(3,3,3,3,3), top = TRUE)%>%align( align = "center", part = "all")%>%bg(bg="#edd77e", j=c(4,5,6,10,11,12))%>%fit_to_width(max_width = 8.25)
} else {
  ft_runinfo=autofit(flextable(trimtable)%>%set_caption(caption = "Table 5: Adaptor and Quality Trimming Summary")%>% theme_zebra()%>%bg(bg = "#44B29C", part = "header"))%>%align( align = "center", part = "all")%>%fit_to_width(max_width = 8.25)
}
flextable_to_rmd(ft_runinfo)
write.csv(trimtable, file=paste0(csvout, "/table5.csv"))
```

```{r Figures2a1, echo=FALSE,fig.align='center', message=FALSE, results='asis'}
###Add support for unaligned_1/unaligned_2 from mapping file info
###Nov 15 2022
flag=FALSE
pmapf<-mapf[mapf$project==projid,]
pmapsamples<-pmapf$sample_id
fig_map<-pmapf
fig_map$demux_file<-gsub("Reports.*", "", fig_map$demux_file)
fig_map$fig_path<-paste0(fig_map$demux_file,"QA/Project_", projid, "/Sample_", fig_map$sample_id)
samples=pmapf$sample_id
bq=list()
testimg=fig_map$fig_path[1]
fqcdir=list.files(path=testimg, pattern=".png", recursive=T, full.names = T)
pbqc1=fqcdir[grepl("per_base_quality.png",fqcdir)]
pdqc=pbqc1[!grepl("_trimmed", pbqc1)]
if(length(pdqc)>1){
  flag=TRUE
}

if(flag==TRUE){
  str1="Per Sample FastQC Summary Results"
  cat(paste0("# ",str1,"\n"))

####

for(p in fig_map$fig_path){
  fqcdir=list.files(path=p, pattern=".png", recursive=T, full.names = T)
  pbqc1=fqcdir[grepl("per_base_quality.png",fqcdir)]
  pdqc=pbqc1[!grepl("_trimmed", pbqc1)]
  bq=append(bq,pdqc)
  #print(pdqc)
  images<-image_read(pdqc)
  #if(images[file.size(images) != 0L]){
  first1<-grid::rasterGrob(as.raster(images[1]))
  first2<-grid::rasterGrob(as.raster(images[2]))
  #cap<-paste0("per base quality for ", s)
  ss=fig_map[fig_map$fig_path==p,]
  s=ss$sample_id
  cap1<-paste0("Read 1 per base quality for ", s)
  cap2<-paste0("Read 2 per base quality for ", s)
  grid.arrange(first1, top = textGrob(cap1, gp = gpar(fontface = 3L, fontsize = 5)))
  grid.arrange(first2, top = textGrob(cap2, gp = gpar(fontface = 3L, fontsize = 5)))
}
basequal_list=unlist(bq)
}
```

```{r rawqc, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", results='asis'}
#runinfo sometimes has 2 lines before relevant text. Normally just one, old runs might have 2.

Str1="Raw Data QC - Contamination Screen Results"

Str2="For each library, a randomly selected subset of reads are screened for contamination by MegaBLAST against a local copy of the NCBI nucleotide (nt) database. The purpose of this screen is only to identify potential contamination in the sample, not to fully characterize the resulting data set."

##edit below for conditional printing borrowed from pacBio
pmapf<-mapf[mapf$project==projid,]
pmapsamples<-pmapf$sample_id
MB_map<-pmapf
MB_map$demux_file<-gsub("Reports.*", "", MB_map$demux_file)
MB_map$mb_path<-paste0(MB_map$demux_file,"QC/Project_", projid, "/Sample_", MB_map$sample_id, "/QC")
samples=pmapf$sample_id
flag=FALSE

###Neater implementation possible for flag assignment. Added here to reduce lines
for(p in MB_map$mb_path){
  mgb=unlist(list.files(path=p,pattern=".fasta.blast.topHits.txt", full.names = T))
    if(file.exists(mgb) && length(mgb)!=0){
      flag=TRUE
    }
}
if(flag==TRUE){
  cat(paste0("# ",Str1,"\n"))
  cat(paste0("## ",Str2))
}
```


```{r Megablast, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center",fig.width=22, fig.height=18}
###Add support for unaligned_1/unaligned_2 from mapping file info
###Nov 15 2022
pmapf<-mapf[mapf$project==projid,]
pmapsamples<-pmapf$sample_id
#samples=tablef$Sample
#pmapf<-mapf[mapf$project==projid,]
#pmapsamples<-pmapf$sample_id
MB_map<-pmapf
MB_map$demux_file<-gsub("Reports.*", "", MB_map$demux_file)
MB_map$mb_path<-paste0(MB_map$demux_file,"QC/Project_", projid, "/Sample_", MB_map$sample_id, "/QC")
samples=pmapf$sample_id

##Deal with multiple lanes for one sample 
map_nodupes<-MB_map[!duplicated(MB_map$sample_id),]

###
for(p in map_nodupes$mb_path){
    #mgb1=paste0(runpath,"/Unaligned/QC/Project_", projid, "/Sample_",s,"/QC/")
    mgb=unlist(list.files(path=p,pattern=".fasta.blast.topHits.txt", full.names = T))
    if(length(mgb)>1){
      mgb=mgb[1]
    }
    if(file.exists(mgb) && length(mgb)!=0){
    text4=readLines(mgb)
    chunksplit <- which(!nzchar(text4))
    if(identical(chunksplit, integer(0))==F){
    start1=chunksplit[1]+1
    end1=chunksplit[2]-1
    chunk1=text4[start1:end1]
    t4p1=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk1)), header=F, sep="\t", row.names = NULL, check.names=FALSE)
    samplename=basename(mgb)
    sname1=gsub( ".fasta.blast.topHits.txt","", samplename)
    sname=gsub(".subset","",sname1)
    names(t4p1)<-NULL
    ###make table grob
    t4_f1<-tableGrob(t4p1, theme=mytheme,rows=NULL)
    h1 <- grobHeight(t4_f1)
    w1 <- grobWidth(t4_f1)
    title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust=0)
    gt1 <- gTree(children=gList(t4_f1,title1))

    start2=end1+2
    end2=chunksplit[3]-1
    chunk2=text4[start2:end2]
    t4p2=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk2)), header=F, sep="\t", row.names = NULL, check.names=FALSE)
    names(t4p2)<-NULL
    ###make table grob
    t4_f2<-tableGrob(t4p2, theme=mytheme,rows=NULL)
    genus=sapply(strsplit(text4[1], ":"), tail, 1)
    species=trimws(sapply(strsplit(text4[2], ":"), tail, 1))
    table4title1=strwrap(paste0("Table 6: Summary Results for ",sname,". This table lists metrics for the expected genus ",genus," & species ", species,"."),width = 175, simplify = F)
    title_wrapped<-sapply(table4title1, paste, collapse = "\n")
    h1 <- grobHeight(t4_f2)
    w1 <- grobWidth(t4_f2)
    title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust=0)
    gtt1<-arrangeGrob(t4_f1,t4_f2, ncol=2)
    gt2 <- gTree(children=gList(gtt1,title1))

    start3=chunksplit[3]+3
    chunk3=text4[start3:length(text4)]
    t4p3=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk3)), header=T, sep="\t", row.names = NULL, check.names=FALSE)
    t5<-t4p3[1:25,]
    t5p1<-t5[rowSums(is.na(t5)) != ncol(t5),]
    ###make table grob
    t4_f3<-tableGrob(t5p1, theme=mytheme,rows=NULL)
    table4title1=strwrap(paste0("Table 7: Contamination Screen Summary for ",sname,". This table lists the top 25 matches for sample the library sequenced, ranked by the number of reads for each hit."),width = 175, simplify = F)
    title_wrapped<-sapply(table4title1, paste, collapse = "\n")
    h1 <- grobHeight(t4_f3)
    w1 <- grobWidth(t4_f3)
    title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust=0)
    #gt1 <- gTree(children=gList(t4_f3,title1))
    gt1=autofit(flextable(t4p3[1:25,])%>%set_caption(caption = paste0("Table 7: Contamination Screen Summary for ",sname,". This table lists the top 25 matches for sample the library sequenced, ranked by the number of reads for each hit."))%>%theme_zebra()%>%bg(bg = "#44B29C", part = "header"))%>% as_raster()
    gg1=ggplot() + theme_void() + annotation_custom(rasterGrob(gt1), xmin=-Inf, xmax=Inf, ymin=-Inf, ymax=Inf)
    grid.arrange(arrangeGrob(gt2), 
             arrangeGrob(gg1), 
             nrow=2)
    write.csv(t4p1, file=paste0(csvout, "/table6a.csv"))
    write.csv(t4p2, file=paste0(csvout, "/table6b.csv"))
    write.csv(t4p3, file=paste0(csvout, "/table7.csv"))
}
}}

```
