---
title: "Maryland Genomics Illumina Report"
output: pdf_document
geometry: margin=0.5cm
params:
  projectname: "Project"
  thedate: !r Sys.Date()
  pathd: "/"
  projid: "A01_1"
  runid: "20220313_S64049_1"
---

```{r setup, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
require(XML)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE)
#classoption: landscape
mytheme <- gridExtra::ttheme_default(base_size=18, core = list(fg_params=list(cex = 0.75, fontface="plain")),colhead = list(fg_params=list(cex = 0.75, fontface="bold")),rowhead = list(fg_params=list(cex = 0.75, fontface="bold")))
mytheme1 <- gridExtra::ttheme_default(base_size=18, core = list(fg_params=list(cex = 0.75, fontface="plain")),colhead = list(fg_params=list(cex = 0.75, fontface="bold")),rowhead = list(fg_params=list(cex = 0.75, fontface="bold")))
mytheme2 <- gridExtra::ttheme_default(base_size=15, core = list(fg_params=list(cex = 0.75, fontface="plain")),colhead = list(fg_params=list(cex = 0.75, fontface="bold")),rowhead = list(fg_params=list(cex = 0.75, fontface="bold")))
library(gridExtra)
library(dplyr)
library(grid)
library(cowplot)
library(magick)
library(ggplot2)
odirec=unlist(params$pathd)
#runid="220418_A00904_0224_AHJ2MKDSX3"
runid=unlist(params$runid)
projid=unlist(params$projid)
#cellid=unlist(params$cellid)
basep="/local/sequence/illumina/data/2022/"
inst_p="/local/projects/grc/pipelines_conf/instruments.conf"
runpath=paste0(basep,runid)
r_param=paste0(runpath, "/RunParameters.xml")
qcplots=paste0(runpath,"/PA_Logs/")
```

```{r logo, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}
knitr::include_graphics("/local/projects-t3/achatterjee/MDG/md-genomics-horizontal.pdf")
#knitr::include_graphics("/Users/apaala.chatterjee/MDG/md-genomics-horizontal.jpeg")
#classoption: landscape
```
This report contains run information and quality metrics for an Illumina run generated by Maryland Genomics. If the project requires
multiple flowcell runs, a separate report will be generated for each flowcell. For additional information or questions, please
contact IGS-Services@SOM.UMaryland.edu

```{r sampledat, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center",fig.width=22, fig.height=14}
#runinf="/Users/apaala.chatterjee/MDG/Illumina/runinfoex2.csv"
runinf=paste0(qcplots,"/runinfo.csv")

sampletab=read.csv(runinf, skip=1, header = T, row.names = NULL)
cnames=colnames(sampletab)[2:length(colnames(sampletab))]
sampletab$index_sequence=NULL
colnames(sampletab)=cnames
proj_samples=sampletab[sampletab$project==projid,]
lanes=unlist(unique(proj_samples$lane))
      if(length(lanes)>1){
        lane=paste0(lanes, collapse = ",")
      } else {
        lane=toString(lanes)
      }
proj_samples$sequencing_sub_sample_id<-NULL
proj_samples$loaded<-NULL
proj_samples$needed<-NULL
proj_samples$mp_insert_size<-NULL
proj_samples$median_smear_size<-NULL
# Reorg columns and change colnames
#1,9,2, 3,10, 5, 7,8,6,11,12,4,13

cnames=c("Lane","Project","Master Sample ID", "Library ID", "Sample Name","External ID","Type","Species","Strain","Method","Index Type","Index Number","Index Sequence")

proj_samplesf=proj_samples[,c(1,9,2,3,10,5,7,8,6,11,12,4,13)]
colnames(proj_samplesf)<-cnames
if(all(is.na(proj_samplesf$Strain))){
  proj_samplesf$Strain=NULL
}
proj_samplesff<-proj_samplesf[rowSums(is.na(proj_samplesf)) != ncol(proj_samplesf),]
proj_samplesf<-proj_samplesff

### Logic to break up long tables 
if(nrow(proj_samplesf)<=35){
  #one
  t1<-tableGrob(proj_samplesf, theme=mytheme1, rows=NULL)
  table1title=strwrap(paste0("Table 1: Sample summary table for samples from project ",projid),width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  
  } else if(nrow(proj_samplesf)>35 && nrow(proj_samplesf)<=70 ){
  #two
  split_point<-35
  df1<-proj_samplesf[1:split_point,]
  df2<-proj_samplesf[(split_point+1):nrow(proj_samplesf),]
  df1<-df1[rowSums(is.na(df1)) != ncol(df1),]
  df2<-df2[rowSums(is.na(df2)) != ncol(df2),]
  
  ###Make double tables
  ##table1 
  t1<-tableGrob(df1, theme=mytheme1, rows=NULL)
  table1title=strwrap(paste0("Table 1: Sample summary table for samples from project ",projid),width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  ##table2
  t2<-tableGrob(df2, theme=mytheme1, rows=NULL)
  table1title=strwrap(paste0("Table 1: Sample summary table for samples from project ",projid),width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t2)
  w1 <- grobWidth(t2)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt2 <- gTree(children=gList(t2, title1))
  grid.arrange(gt2)
  
} else if(nrow(proj_samplesf)>70 && nrow(proj_samplesf)<=105){
  #three
  #split_lines1<-round(nrow(proj_samplesf)/3)
  split_lines1<-35
  split_lines2<-split_lines1*2
  df1<-proj_samplesf[1:split_lines1,]
  df2<-proj_samplesf[(split_lines1+1):split_lines2,]
  df3<-proj_samplesf[(split_lines2+1):nrow(proj_samplesf),]
  
  df1<-df1[rowSums(is.na(df1)) != ncol(df1),]
  df2<-df2[rowSums(is.na(df2)) != ncol(df2),]
  df3<-df3[rowSums(is.na(df3)) != ncol(df3),]
  
  #table1
  t1<-tableGrob(df1, theme=mytheme1, rows=NULL)
  table1title=strwrap(paste0("Table 1: Sample summary table for samples from project ",projid),width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  ##table2
  t2<-tableGrob(df2, theme=mytheme1, rows=NULL)
  table1title=strwrap(paste0("Table 1: Sample summary table for samples from project ",projid),width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t2)
  w1 <- grobWidth(t2)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt2 <- gTree(children=gList(t2, title1))
  grid.arrange(gt2)
  ##table 3
  t3<-tableGrob(df3, theme=mytheme1, rows=NULL)
  table1title=strwrap(paste0("Table 1: Sample summary table for samples from project ",projid),width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t3)
  w1 <- grobWidth(t3)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt3 <- gTree(children=gList(t3, title1))
  grid.arrange(gt3)
}else if(nrow(proj_samplesf)>105 && nrow(proj_samplesf)<=140){
  #four
  split_lines1<-round(nrow(proj_samplesf)/4)
  split_lines2<-split_lines1*2
  split_lines3<-split_lines1*3
  
  split_lines1<-40
  split_lines2<-split_lines1*2
  split_lines3<-split_lines1*3
  
  df1<-proj_samplesf[1:split_lines1,]
  df2<-proj_samplesf[(split_lines1+1):split_lines2,]
  df3<-proj_samplesf[(split_lines2+1):split_lines3,]
  df4<-proj_samplesf[(split_lines3+1):nrow(proj_samplesf),]
  
  df1<-df1[rowSums(is.na(df1)) != ncol(df1),]
  df2<-df2[rowSums(is.na(df2)) != ncol(df2),]
  df3<-df3[rowSums(is.na(df3)) != ncol(df3),]
  df4<-df4[rowSums(is.na(df4)) != ncol(df4),]
  #table1
  t1<-tableGrob(df1, theme=mytheme1, rows=NULL)
  table1title=strwrap(paste0("Table 1: Sample summary table for samples from project ",projid),width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  ##table2
  t2<-tableGrob(df2, theme=mytheme1, rows=NULL)
  table1title=strwrap(paste0("Table 1: Sample summary table for samples from project ",projid),width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t2)
  w1 <- grobWidth(t2)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt2 <- gTree(children=gList(t2, title1))
  grid.arrange(gt2)
  ##table 3
  t3<-tableGrob(df3, theme=mytheme1, rows=NULL)
  table1title=strwrap(paste0("Table 1: Sample summary table for samples from project ",projid),width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t3)
  w1 <- grobWidth(t3)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt3 <- gTree(children=gList(t3, title1))
  grid.arrange(gt3)
    ##table 4
  t4<-tableGrob(df4, theme=mytheme1, rows=NULL)
  table1title=strwrap(paste0("Table 1: Sample summary table for samples from project ",projid),width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t4)
  w1 <- grobWidth(t4)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt4 <- gTree(children=gList(t4, title1))
  grid.arrange(gt4)
}


```

```{r rinfo, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}
#rif="/Users/apaala.chatterjee/MDG/Illumina/RunInfo.xml"
rif=paste0(runpath,"/RunInfo.xml")
data <- xmlParse(rif)
xml_data <- xmlToList(data)
recipe_xml=xml_data$Run$Reads
recipe_t=""
for(r in recipe_xml){
  recipe_t=paste0(recipe_t,":", r[2])
}
recipe=substring(recipe_t, 2)

#classoption: landscape

###From Ivette
instrument_code=unlist(strsplit(runid, "_"))[2]
inst_look=read.csv(inst_p, comment.char = "#", header=F)
hold=inst_look[inst_look$V2==instrument_code,]
Inst_type=hold$V3

dat <- xmlParse(r_param)
xml_data <- xmlToList(dat)
flowcell_type=xml_data$RfidsInfo$FlowCellMode
hdat=data.frame("InstrumentType"=Inst_type, "FlowCellType"=flowcell_type, "RunRecipe"=recipe)
t1<-tableGrob(hdat, theme=mytheme2, rows=NULL)
table1title=strwrap("Table 2:  Run info, including instrument type, run & flowcell type",width = 175, simplify = F)
title_wrapped<-sapply(table1title, paste, collapse = "\n")
h1 <- grobHeight(t1)
w1 <- grobWidth(t1)
title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
gt1 <- gTree(children=gList(t1, title1))

grid.arrange(gt1)
```

```{r Figures1a, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="Flowcell Passed-Filter Cluster Count Distribution by Lane."}
newplots=paste0(qcplots,"/images/")
if(length(lanes)==1){
fig1n=unlist(list.files(path=newplots, pattern=paste0(lanes,"_ClusterCount-by-lane.png"), full.names = T))
fig1 = fig1n[file.size(fig1n) != 0L]
if(file.exists(fig1)){
#if(file.size(fig1) != 0L){
ggdraw() + draw_image(fig1)
}}

if(length(lanes)>1){
  fig1n=unlist(list.files(path=newplots, pattern="all_ClusterCount-by-lane.png", full.names = T))
fig1 = fig1n[file.size(fig1n) != 0L]
if(file.exists(fig1)){
#if(file.size(fig1) != 0L){
ggdraw() + draw_image(fig1)
}}
```

```{r Figures1b, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="Flowcell Signal Intensity By Lane"}
if(length(lanes)==1){
fig2n=unlist(list.files(path=newplots, pattern=paste0(lanes,"_flowcell-Intensity.png"), full.names = T))
fig2 = fig2n[file.size(fig2n) != 0L]
if(file.exists(fig2)){
#if(file.size(fig1) != 0L){
ggdraw() + draw_image(fig2)
}}

if(length(lanes)>1){
  fig2n=unlist(list.files(path=newplots, pattern="all_flowcell-Intensity.png", full.names = T))
fig2 = fig1n[file.size(fig2n) != 0L]
if(file.exists(fig2)){
#if(file.size(fig1) != 0L){
ggdraw() + draw_image(fig2)
}}
```

```{r Figures1c, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="Flowcell Signal Intensity By Cycle"}
if(length(lanes)==1){
fig3n=unlist(list.files(path=newplots, pattern=paste0(lanes,"_Intensity-by-cycle_Intensity.png"), full.names = T))
fig3 = fig3n[file.size(fig3n) != 0L]
if(file.exists(fig3)){
#if(file.size(fig1) != 0L){
ggdraw() + draw_image(fig3)
}}

if(length(lanes)>1){
  fig3n=unlist(list.files(path=newplots, pattern="all_Intensity-by-cycle_Intensity.png", full.names = T))
fig3 = fig3n[file.size(fig3n) != 0L]
if(file.exists(fig3)){
#if(file.size(fig1) != 0L){
ggdraw() + draw_image(fig3)
}}
```

```{r Figures1d, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="Flowcell Mean Quality Score Distribution By Cycle"}
if(length(lanes)==1){
fig4n=unlist(list.files(path=newplots, pattern=paste0(lanes,"_q-heat-map.png"), full.names = T))
fig4 = fig4n[file.size(fig4n) != 0L]
if(file.exists(fig4)){
#if(file.size(fig1) != 0L){
ggdraw() + draw_image(fig4)
}}

if(length(lanes)>1){
  fig4n=unlist(list.files(path=newplots, pattern="all_q-heat-map.png", full.names = T))
fig4 = fig4n[file.size(fig4n) != 0L]
if(file.exists(fig4)){
#if(file.size(fig1) != 0L){
ggdraw() + draw_image(fig4)
}}
```


```{r qcmetrics, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}
summary=paste0(qcplots,"/summary.csv")
#summary="/Users/apaala.chatterjee/MDG/Illumina/summary.csv"
sumtext=readLines(summary)
i1=which(!nzchar(sumtext)) [1] 
i1=i1-1
readsum=sumtext[3:i1]
#gsub("\"","",readsum)
summary_table=read.table(text=readsum,header=T, fill = T, as.is = T, sep=",", check.names = F)

t1<-tableGrob(summary_table, theme=mytheme2, rows=NULL)
table1title=strwrap("Table 3: Flowcell QC Metrics",width = 175, simplify = F)
title_wrapped<-sapply(table1title, paste, collapse = "\n")
h1 <- grobHeight(t1)
w1 <- grobWidth(t1)
title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
gt1 <- gTree(children=gList(t1, title1))

grid.arrange(gt1)
```


```{r demux1, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}
##mapping file implementstion
demux_map<-paste0(qcplots, "demux_map_table.csv")
mapf<-read.table(demux_map, sep=",", header=T)
pmaps<-mapf[mapf$project==projid,]
t1_list<-data.frame()
t2_list<-data.frame()
t1f=FALSE
t2f=FALSE
t3f=FALSE
samples<-proj_samples$sample_id
smapf<-mapf[mapf$sample_id %in% samples,]

for(path in smapf$demux_file){
  alltables=readHTMLTable(path)
  for(i in alltables){
  if(is.null(i)==F){
    if("Project" %in% colnames(i)){
      hold=i[rowSums(is.na(i)) != ncol(i),]
      proj=paste0("Project_",projid)
      tablef=hold[hold$Project==proj,]
      tablef$`PF Clusters`<-as.numeric(gsub(",","",tablef$`PF Clusters`))
      tablef$`Yield (Mbases)`<-as.numeric(gsub(",","",tablef$`Yield (Mbases)`))
      t1_list<-rbind(t1_list,tablef)
      yield=sum(tablef$`Yield (Mbases)`)
      lanes=unlist(unique(tablef$Lane))
      if(length(lanes)>1){
        lane=paste0(lanes, collapse = ",")
      } else {
        lane=toString(lanes)
      }
      #lane=paste(shQuote(lanes), collapse=", ")
      pfcluster=sum(tablef$`PF Clusters`)
      tab1=data.frame("Clusters(PF)"=prettyNum(pfcluster,big.mark=",",scientific=FALSE), "Yield (MBases)"=prettyNum(yield,big.mark=",",scientific=FALSE), check.names = F)
      #tab1<-prettyNum(tab11, big.mark=",",scientific=FALSE)
    }}}}
tablef<-unique(t1_list)

pfcluster=sum(tablef$`PF Clusters`)
      tab1=data.frame("Clusters(PF)"=prettyNum(pfcluster,big.mark=",",scientific=FALSE), "Yield (MBases)"=prettyNum(yield,big.mark=",",scientific=FALSE), check.names = F)
      t1<-tableGrob(tab1, theme=mytheme2, rows=NULL)
      table1title=strwrap(paste0("Table 4a: Total Passed-Filter Data Generated for Project ", projid),width = 175, simplify = F)
      title_wrapped<-sapply(table1title, paste, collapse = "\n")
      h1 <- grobHeight(t1)
      w1 <- grobWidth(t1)
      title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
      t1f <- gTree(children=gList(t1, title1))
      ##edit table content.
      tablef$Project<-gsub("Project_","", tablef$Project)
      tablef$`Barcode sequence`<-NULL
      #Add summary row at bottom
      #Add summary row with sums of PF Clusters, % of the lane, and Yield. And averages of the remaining columns
      tablef[c(4,5,6,7,8,9,10,11)]<-sapply(tablef[c(4,5,6,7,8,9,10,11)],as.numeric)
      table4<-tablef %>% bind_rows(summarise(.,
                        across(where(is.numeric), sum),
                        across(where(is.character), ~"Total")))
      nums=nrow(tablef)
      lastr=nums+1
      table4$`Mean QualityScore`[lastr]<-table4$`Mean QualityScore`[lastr]/nums
      table4$`% Perfectbarcode`[lastr]<-table4$`% Perfectbarcode`[lastr]/nums
      table4$`% One mismatchbarcode`[lastr]<-table4$`% One mismatchbarcode`[lastr]/nums
      table4$`% PFClusters`[lastr]<-table4$`% PFClusters`[lastr]/nums
      table4$`% >= Q30bases`[lastr]<-table4$`% >= Q30bases`[lastr]/nums
      t4_sampleID<-merge(table4, proj_samplesf[, c("Library ID", "Master Sample ID")], by.x="Sample", by.y = "Library ID", all.x=T)
      table4f<-t4_sampleID[,c(1,12,2,3,4,5,6,7,8,9,10,11)]
      names(table4f)[names(table4f) == 'Sample'] <- 'Library ID'
      names(table4f)[names(table4f) == 'Master Sample ID'] <- 'Sample ID'
      table4f$`Sample ID`[nrow(table4f)]<-"Total"
      t1<-tableGrob(table4f, theme=mytheme1, rows=NULL)
      table1title=strwrap("Table 4b: Per-Sample Summary Statistics",width = 175, simplify = F)
      title_wrapped<-sapply(table1title, paste, collapse = "\n")
      h1 <- grobHeight(t1)
      w1 <- grobWidth(t1)
      title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
      t2f <- gTree(children=gList(t1, title1))
grid.arrange(t1f)
```

```{r demux2, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center",fig.width=20, fig.height=14}
####split table
if(nrow(tablef)<=35){
  #one
  t1<-tableGrob(tablef, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 4b: Per-Sample Summary Statistics",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  
  } else if(nrow(tablef)>35 && nrow(tablef)<=70 ){
  #two
  split_point<-35
  df1<-tablef[1:split_point,]
  df2<-tablef[(split_point+1):nrow(tablef),]
  df1<-df1[rowSums(is.na(df1)) != ncol(df1),]
  df2<-df2[rowSums(is.na(df2)) != ncol(df2),]
  
  ###Make double tables
  ##table1 
  t1<-tableGrob(df1, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 4b: Per-Sample Summary Statistics",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  ##table2
  t2<-tableGrob(df2, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 4b: Per-Sample Summary Statistics",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t2)
  w1 <- grobWidth(t2)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt2 <- gTree(children=gList(t2, title1))
  grid.arrange(gt2)
  
} else if(nrow(tablef)>70 && nrow(tablef)<=105){
  #three
  #split_lines1<-round(nrow(tablef)/3)
  split_lines1<-35
  split_lines2<-split_lines1*2
  df1<-tablef[1:split_lines1,]
  df2<-tablef[(split_lines1+1):split_lines2,]
  df3<-tablef[(split_lines2+1):nrow(tablef),]
  
  df1<-df1[rowSums(is.na(df1)) != ncol(df1),]
  df2<-df2[rowSums(is.na(df2)) != ncol(df2),]
  df3<-df3[rowSums(is.na(df3)) != ncol(df3),]
  
  #table1
  t1<-tableGrob(df1, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 4b: Per-Sample Summary Statistics",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  ##table2
  t2<-tableGrob(df2, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 4b: Per-Sample Summary Statistics",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t2)
  w1 <- grobWidth(t2)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt2 <- gTree(children=gList(t2, title1))
  grid.arrange(gt2)
  ##table 3
  t3<-tableGrob(df3, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 4b: Per-Sample Summary Statistics",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t3)
  w1 <- grobWidth(t3)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt3 <- gTree(children=gList(t3, title1))
  grid.arrange(gt3)
}else if(nrow(tablef)>105 && nrow(tablef)<=140){
  #four
  split_lines1<-round(nrow(tablef)/4)
  split_lines2<-split_lines1*2
  split_lines3<-split_lines1*3
  
  split_lines1<-40
  split_lines2<-split_lines1*2
  split_lines3<-split_lines1*3
  
  df1<-tablef[1:split_lines1,]
  df2<-tablef[(split_lines1+1):split_lines2,]
  df3<-tablef[(split_lines2+1):split_lines3,]
  df4<-tablef[(split_lines3+1):nrow(tablef),]
  
  df1<-df1[rowSums(is.na(df1)) != ncol(df1),]
  df2<-df2[rowSums(is.na(df2)) != ncol(df2),]
  df3<-df3[rowSums(is.na(df3)) != ncol(df3),]
  df4<-df4[rowSums(is.na(df4)) != ncol(df4),]
  #table1
  t1<-tableGrob(df1, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 4b: Per-Sample Summary Statistics",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  ##table2
  t2<-tableGrob(df2, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 4b: Per-Sample Summary Statistics",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t2)
  w1 <- grobWidth(t2)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt2 <- gTree(children=gList(t2, title1))
  grid.arrange(gt2)
  ##table 3
  t3<-tableGrob(df3, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 4b: Per-Sample Summary Statistics",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t3)
  w1 <- grobWidth(t3)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt3 <- gTree(children=gList(t3, title1))
  grid.arrange(gt3)
    ##table 4
  t4<-tableGrob(df4, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 4b: Per-Sample Summary Statistics",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t4)
  w1 <- grobWidth(t4)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt4 <- gTree(children=gList(t4, title1))
  grid.arrange(gt4)
}
```

```{r trimstats, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center",fig.width=20, fig.height=14}
#tstats="/Users/apaala.chatterjee/MDG/Illumina/trimming.logs.csv"
tstats=paste0(qcplots, "/trimming.logs.csv")
trimstats=read.csv(tstats, header = T, row.names = NULL)
trimf=trimstats[trimstats$Project==projid,]
trimf$Lane<-gsub("L00","",trimf$Lane)

####split table
if(nrow(trimf)<=35){
  #one
  t1<-tableGrob(trimf, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 5: Adaptor and Quality Trimming Summary",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  
  } else if(nrow(trimf)>35 && nrow(trimf)<=70 ){
  #two
  split_point<-35
  df1<-trimf[1:split_point,]
  df2<-trimf[(split_point+1):nrow(trimf),]
  df1<-df1[rowSums(is.na(df1)) != ncol(df1),]
  df2<-df2[rowSums(is.na(df2)) != ncol(df2),]
  
  ###Make double tables
  ##table1 
  t1<-tableGrob(df1, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 5: Adaptor and Quality Trimming Summary",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  ##table2
  t2<-tableGrob(df2, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 5: Adaptor and Quality Trimming Summary",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t2)
  w1 <- grobWidth(t2)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt2 <- gTree(children=gList(t2, title1))
  grid.arrange(gt2)
  
} else if(nrow(trimf)>70 && nrow(trimf)<=105){
  #three
  #split_lines1<-round(nrow(trimf)/3)
  split_lines1<-35
  split_lines2<-split_lines1*2
  df1<-trimf[1:split_lines1,]
  df2<-trimf[(split_lines1+1):split_lines2,]
  df3<-trimf[(split_lines2+1):nrow(trimf),]
  
  df1<-df1[rowSums(is.na(df1)) != ncol(df1),]
  df2<-df2[rowSums(is.na(df2)) != ncol(df2),]
  df3<-df3[rowSums(is.na(df3)) != ncol(df3),]
  
  #table1
  t1<-tableGrob(df1, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 5: Adaptor and Quality Trimming Summary",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  ##table2
  t2<-tableGrob(df2, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 5: Adaptor and Quality Trimming Summary",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t2)
  w1 <- grobWidth(t2)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt2 <- gTree(children=gList(t2, title1))
  grid.arrange(gt2)
  ##table 3
  t3<-tableGrob(df3, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 5: Adaptor and Quality Trimming Summary",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t3)
  w1 <- grobWidth(t3)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt3 <- gTree(children=gList(t3, title1))
  grid.arrange(gt3)
}else if(nrow(trimf)>105 && nrow(trimf)<=140){
  #four
  split_lines1<-round(nrow(trimf)/4)
  split_lines2<-split_lines1*2
  split_lines3<-split_lines1*3
  
  split_lines1<-40
  split_lines2<-split_lines1*2
  split_lines3<-split_lines1*3
  
  df1<-trimf[1:split_lines1,]
  df2<-trimf[(split_lines1+1):split_lines2,]
  df3<-trimf[(split_lines2+1):split_lines3,]
  df4<-trimf[(split_lines3+1):nrow(trimf),]
  
  df1<-df1[rowSums(is.na(df1)) != ncol(df1),]
  df2<-df2[rowSums(is.na(df2)) != ncol(df2),]
  df3<-df3[rowSums(is.na(df3)) != ncol(df3),]
  df4<-df4[rowSums(is.na(df4)) != ncol(df4),]
  #table1
  t1<-tableGrob(df1, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 5: Adaptor and Quality Trimming Summary",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t1)
  w1 <- grobWidth(t1)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt1 <- gTree(children=gList(t1, title1))
  grid.arrange(gt1)
  ##table2
  t2<-tableGrob(df2, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 5: Adaptor and Quality Trimming Summary",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t2)
  w1 <- grobWidth(t2)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt2 <- gTree(children=gList(t2, title1))
  grid.arrange(gt2)
  ##table 3
  t3<-tableGrob(df3, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 5: Adaptor and Quality Trimming Summary",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t3)
  w1 <- grobWidth(t3)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt3 <- gTree(children=gList(t3, title1))
  grid.arrange(gt3)
    ##table 4
  t4<-tableGrob(df4, theme=mytheme1, rows=NULL)
  table1title=strwrap("Table 5: Adaptor and Quality Trimming Summary",width = 175, simplify = F)
  title_wrapped<-sapply(table1title, paste, collapse = "\n")
  h1 <- grobHeight(t4)
  w1 <- grobWidth(t4)
  title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust = 0)
  gt4 <- gTree(children=gList(t4, title1))
  grid.arrange(gt4)
}


```
# Per Sample FastQC Summary Results

```{r Figures2a1, echo=FALSE,fig.align='center', message=FALSE}

samples=tablef$Sample
bq=list()
for(s in samples){
fqc1=paste0(runpath,"/Unaligned/QA/Project_", projid, "/Sample_",s)
fqcdir=list.files(path=fqc1, pattern=".png", recursive=T, full.names = T)
pbqc1=fqcdir[grepl("per_base_quality.png",fqcdir)]
pdqc=pbqc1[!grepl("_trimmed", pbqc1)]
bq=append(bq,pdqc)
images<-image_read(pdqc)
#if(images[file.size(images) != 0L]){
first1<-grid::rasterGrob(as.raster(images[1]))
first2<-grid::rasterGrob(as.raster(images[2]))
#cap<-paste0("per base quality for ", s)
cap1<-paste0("Read 1 per base quality for ", s)
cap2<-paste0("Read 2 per base quality for ", s)
grid.arrange(first1, top = textGrob(cap1, gp = gpar(fontface = 3L, fontsize = 5)))
grid.arrange(first2, top = textGrob(cap2, gp = gpar(fontface = 3L, fontsize = 5)))
}
#}
basequal_list=unlist(bq)
```

# Raw Data QC - Contamination Screen Results

## For each library, a randomly selected subset of reads are screened for contamination by MegaBLAST against a local copy of the NCBI nucleotide (nt) database. The purpose of this screen is only to identify potential contamination in the sample, not to fully characterize the resulting data set.


```{r Megablast, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center",fig.width=22, fig.height=18}

samples=tablef$Sample
for(s in samples){
    mgb1=paste0(runpath,"/Unaligned/QC/Project_", projid, "/Sample_",s,"/QC/")
    mgb=unlist(list.files(path=mgb1,pattern="subset.fasta.blast.topHits.txt", full.names = T))
    if(file.exists(mgb)){
    text4=readLines(mgb)
    chunksplit <- which(!nzchar(text4))
    if(identical(chunksplit, integer(0))==F){
    start1=chunksplit[1]+1
    end1=chunksplit[2]-1
    chunk1=text4[start1:end1]
    t4p1=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk1)), header=F, sep="\t", row.names = NULL, check.names=FALSE)
    samplename=basename(mgb)
    sname=gsub( ".subset.fasta.blast.topHits.txt","", samplename)
    names(t4p1)<-NULL
    ###make table grob
    t4_f1<-tableGrob(t4p1, theme=mytheme,rows=NULL)
    table4title1=strwrap(paste0("Table 6: Summary Results for ",s,". This table lists metrics for the expected genus & species, and the top other genus and species."),width = 175, simplify = F)
    title_wrapped<-sapply(table4title1, paste, collapse = "\n")
    h1 <- grobHeight(t4_f1)
    w1 <- grobWidth(t4_f1)
    title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust=0)
    gt1 <- gTree(children=gList(t4_f1,title1))

    start2=end1+2
    end2=chunksplit[3]-1
    chunk2=text4[start2:end2]
    t4p2=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk2)), header=F, sep="\t", row.names = NULL, check.names=FALSE)
    names(t4p2)<-NULL
    ###make table grob
    t4_f2<-tableGrob(t4p2, theme=mytheme,rows=NULL)
    table4title1=strwrap(paste0("Table 6: Summary Results for ",s,". This table lists metrics for the expected genus & species, and the top other genus and species."),width = 175, simplify = F)
    title_wrapped<-sapply(table4title1, paste, collapse = "\n")
    h1 <- grobHeight(t4_f2)
    w1 <- grobWidth(t4_f2)
    title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust=0)
    gtt1<-arrangeGrob(t4_f1,t4_f2, ncol=2)
    gt2 <- gTree(children=gList(gtt1,title1))

    start3=chunksplit[3]+3
    chunk3=text4[start3:length(text4)]
    t4p3=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk3)), header=T, sep="\t", row.names = NULL, check.names=FALSE)
    t5<-t4p3[1:25,]
    t5p1<-t5[rowSums(is.na(t5)) != ncol(t5),]
    ###make table grob
    t4_f3<-tableGrob(t5p1, theme=mytheme,rows=NULL)
    table4title1=strwrap(paste0("Table 7: Contamination Screen Summary for ",sname,". This table lists the top 25 matches for sample the library sequenced, ranked by the number of reads for each hit."),width = 175, simplify = F)
    title_wrapped<-sapply(table4title1, paste, collapse = "\n")
    h1 <- grobHeight(t4_f3)
    w1 <- grobWidth(t4_f3)
    title1 <- textGrob(title_wrapped, gp=gpar(fontsize=18), y=unit(0.6,"npc")+ 0.75*h1, vjust=0)
    gt1 <- gTree(children=gList(t4_f3,title1))
    grid.arrange(arrangeGrob(gt2), 
             arrangeGrob(gt1), 
             nrow=2)
}
}}

```
