---
title: "Maryland Genomics PacBio Run Report"
output: 
  pdf_document:
    extra_dependencies: ["float"]
    latex_engine: xelatex
    fig_caption: FALSE
geometry: margin=0.5cm
params:
  projectname: "Project"
  thedate: !r Sys.Date()
  pathd: ""
  cellid: "A01_1"
  runid: "20220313_S64049_1"
  basep: "/local/sequence/pacbio/data/2022/"

header-includes:
   -  \usepackage{fancyhdr}
   -  \pagestyle{fancy}
   -  \usepackage{colortbl} 
   -  \usepackage{color} 
   -  \usepackage{caption}
   -  \usepackage{xcolor} 
   -  \fancyfoot[CO,CE]{PacBio Report V2.1}
---

```{r setup, include=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE)
knitr::opts_chunk$set(fig.pos = 'H')
knitr::opts_chunk$set(tab.cap.pre = "", tab.cap.sep = "")
knitr::opts_chunk$set(echo = FALSE, ft.keepnext = F)
knitr::opts_chunk$set(warning = FALSE, message=FALSE)
```
\captionsetup[table]{labelformat=empty}
```{r setuppaths, echo=FALSE, warning=FALSE, message=FALSE}
library(gridExtra)
library(grid)
library(cowplot)
library(magick)
library(ggplot2)
library(data.table)
library(scales)
library(jsonlite)
library(flextable)
library(webshot)
library(cowplot)
#library(magrittr)
odirec=unlist(params$pathd)
runid=unlist(params$runid)
cellid=unlist(params$cellid)
basep=unlist(params$basep)
basepsun="/PA_Logs/runinfo.csv"
runpath=paste0(basep,runid,basepsun)
statspath=paste0(basep,runid,"/PA_Logs/run_stats.csv")
demuxpath=paste0(basep,runid,"/PA_Logs/demux_stats.txt")
mytheme <- gridExtra::ttheme_default(core = list(fg_params=list(cex = 0.75, fontface="plain")),colhead = list(fg_params=list(cex = 0.75, fontface="bold")),rowhead = list(fg_params=list(cex = 0.75, fontface="bold")))
#make folder for csv tables
csvout=paste0(odirec,"/report_tables")
if (!dir.exists(csvout)){
    dir.create(csvout)
}
```

```{r logo, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center"}
knitr::include_graphics("/local/projects-t3/achatterjee/MDG/md-genomics-horizontal.pdf")
#classoption: landscape
set_flextable_defaults(font.family = 'DejaVuSansMono')
```
This report contains run information and metrics for a PacBio SMRT Cell run generated by Maryland Genomics. If the project requires multiple SMRT Cell runs, a separate report will be generated for each SMRT Cell. For additional information or questions, please contact IGS-Services@SOM.UMaryland.edu

```{r preambletext, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", results='asis'}
#runinfo sometimes has 2 lines before relevant text. Normally just one, old runs might have 2
runinfo=read.csv(runpath,header=TRUE, skip=1,row.names = NULL,fill = T,check.names=FALSE)
colnames(runinfo)=colnames(runinfo)[2:ncol(runinfo)]
runinfo1 <- runinfo[ , - ncol(runinfo)]
runinfo=runinfo1[grepl(cellid, runinfo1$cell),]
movie=unique(runinfo$movies)
moviel=unique(runinfo$movie_length)
cellstat=unique(runinfo$cell_status)
stringpp<-paste0("This report generated for PacBio run ",runid," SMRT Cell ",cellid, ". The number of movies was ", movie,", the length of the movie is ", moviel, " and the cell status is ",cellstat,".")
cat(stringpp)
```


```{r Table1, echo=FALSE, warnings=FALSE,fig.width=16, fig.height=14, message=FALSE, fig.align='center', tab.cap = NULL, results='asis'}
###modify the dataframe to allow duplicate rownames
runinfo=read.csv(runpath, sep=",",header=TRUE, skip=1,row.names = NULL,check.names=FALSE)
colnames(runinfo)=colnames(runinfo)[2:ncol(runinfo)]
runinfo <- runinfo[ , - ncol(runinfo)]
#remove columns indicated by Luke
runinfo$movies=NULL
runinfo$movie_length=NULL
runinfo$cell_status=NULL
#move column for strain next to species
cellruninfo=runinfo[runinfo$cell==paste0(cellid,"_1"),]
if(nrow(cellruninfo)==1){
tcellruninfo=data.frame("Parameter"=colnames(cellruninfo), "Value"=t(cellruninfo), row.names = NULL)
colnames(tcellruninfo)=c("Parameter", "Value")
tcellruninfo[tcellruninfo$Parameter=="insertsize",]["Value"]<-prettyNum(tcellruninfo[tcellruninfo$Parameter=="insertsize",]["Value"],big.mark=",",scientific=FALSE)
tcellruninfo[tcellruninfo$Parameter=="targeted_frg_size",]["Value"]<-prettyNum(tcellruninfo[tcellruninfo$Parameter=="targeted_frg_size",]["Value"],big.mark=",",scientific=FALSE)

if( any(tcellruninfo$Parameter %in% "size_selected_size") ){
tcellruninfo[tcellruninfo$Parameter=="size_selected_size",]["Value"]<-prettyNum(tcellruninfo[tcellruninfo$Parameter=="size_selected_size",]["Value"],big.mark=",",scientific=FALSE)}

#Example using gt package for future refeence
#runinf=gt(tcellruninfo)|>tab_options(column_labels.background.color = "teal")%>%tab_style(style=cell_borders(sides = c("top","bottom")), locations = cells_body(columns = everything(),rows = everything())) %>% tab_header(title = md("Table 1: Sample information. This table lists all samples sequenced on the SMRT Cell."))

###flextable
ft_runinfo=autofit(flextable(tcellruninfo)%>%set_caption(caption = "Table 1: Sample information. This table lists all samples sequenced on the SMRT Cell.")%>% theme_zebra()%>%bg(bg = "#44B29C", part = "header"))
########################
write.csv(tcellruninfo, file=paste0(csvout, "/table1.csv"))
t1<-tableGrob(tcellruninfo, theme=mytheme, rows = NULL)

flextable_to_rmd(ft_runinfo)
} else if(nrow(cellruninfo)>1){
  cellruninfo$insertsize<-prettyNum(cellruninfo$insertsize,big.mark=",",scientific=FALSE)
  cellruninfo$targeted_frg_size<-prettyNum(cellruninfo$targeted_frg_size,big.mark=",",scientific=FALSE)

ft_runinfo=autofit(flextable(cellruninfo)%>%set_caption(caption = "Table 1: Sample information. This table lists all samples sequenced on the SMRT Cell.")%>% theme_zebra()%>%bg(bg = "#44B29C", part = "header")%>%fontsize(size = 12))%>%fit_to_width(max_width = 7.5)
flextable_to_rmd(ft_runinfo)
write.csv(cellruninfo, file=paste0(csvout, "/table1.csv"))
}
```

```{r Table2, echo=FALSE,fig.width=10, fig.height=12, message=FALSE, fig.align='center', warning=FALSE, fig.pos='H', results='asis'}
#Determine number of tables from file
suppressWarnings({
rm(gt1)
rm(gt2)
rm(gt3)
})
l=readLines(statspath)
dropid=gsub("\\_.*","",cellid)
a=grep("PF Polymerase Read Stats:", l)
b=grep("Subreads Stats:", l)
c=grep("CCS Stats:", l)
#Table 1
if(length(a)!=0){
 i1=a+1
 i2=grep("Stats:", l)[2]-1
 if(i2==0){
   i2=length(l)
 }

 t2_p1=read.csv(text= l[i1:i2], header=T, sep=",", check.names=FALSE)
 table1title=strwrap("Table 2A: Polymerase Read Stats. This table contains metrics for all polymerase reads generated by the run.",width = 150, simplify = F)
 title_wrapped<-sapply(table1title, paste, collapse = "\n")
 t2p1=t2_p1[t2_p1$cell==dropid,]
 t2p1[,c(-1,-2,-3)]<-as.numeric(t2p1[,c(-1,-2,-3)])
 for(cc in colnames(t2p1)){
   if (is.numeric(t2p1[,cc])){
     t2p1[,cc]<-prettyNum( t2p1[,cc], big.mark=",",scientific=FALSE)
   }
 }
 #transform table
 trans_t2p1<-data.frame("Parameters"=colnames(t2p1), "Value"=t(t2p1), row.names = NULL)
 colnames(trans_t2p1)=c("Parameters", "Value")
 t2_f1<-tableGrob(trans_t2p1, theme=mytheme,rows=NULL)
 ###flextable
 gt1=autofit(flextable(trans_t2p1)%>%set_caption(caption = "Table 2A: Polymerase Read Stats. This table contains metrics for all polymerase reads generated by the run.")%>% theme_zebra()%>%bg(bg = "#44B29C", part = "header"))
}
#Table 2
if(length(b)!=0){
 i1=b+1
 #i2=grep("Stats:", l[i1:length(l)])[1]
 if(length(c)!=0){
   i2=c-1
 } else
 {
   i2=length(l)
 }
 t2_p2=read.csv(text=l[i1:i2], header=T, sep=",", check.names=FALSE)
 table1title=strwrap("Table 2B: SubRead Stats. This table contains metrics for all subreads generated by the run.",width = 150, simplify = F)
 title_wrapped<-sapply(table1title, paste, collapse = "\n")
 t2p2=t2_p2[t2_p2$cell==dropid,]
 t2p2[,c(-1,-2,-3)]<-as.numeric(t2p2[,c(-1,-2,-3)])
  for(cc in colnames(t2p2)){
   if (is.numeric(t2p2[,cc])){
     t2p2[,cc]<-prettyNum( t2p2[,cc], big.mark=",",scientific=FALSE)
   }
 }
 trans_t2p2<-data.frame("Parameters"=colnames(t2p2), "Value"=t(t2p2), row.names = NULL)
 colnames(trans_t2p2)=c("Parameters", "Value")
 t2_f2<-tableGrob(trans_t2p2, theme=mytheme,rows=NULL)
 ###flextable
 gt2=autofit(flextable(trans_t2p2)%>%set_caption(caption = "Table 2B: SubRead Stats. This table contains metrics for all subreads generated by the run.")%>% theme_zebra()%>%bg(bg = "#44B29C", part = "header"))
}
#Table 3
if(length(c)!=0){
 i1=c+1
 i2=length(l)
 t2_p3=read.csv(text=l[i1:i2], header=T, sep=",", check.names=FALSE)
 table1title=strwrap("Table 2C: CCS/HiFi Stats. This table contains metrics for all HiFi (CCS) reads generated by the run.",width = 150, simplify = F)
 title_wrapped<-sapply(table1title, paste, collapse = "\n")
 t2p3=t2_p3[t2_p3$cell==dropid,]
 if(nrow(t2p3)!=0){
   for(cc in colnames(t2p3)){
   if (is.numeric(t2p3[,cc])){
     t2p3[,cc]<-prettyNum( t2p3[,cc], big.mark=",",scientific=FALSE)
   }
 }
 trans_t2p3<-data.frame("Parameters"=colnames(t2p3), "Value"=t(t2p3), row.names = NULL)
 colnames(trans_t2p3)=c("Parameters", "Value")
 t2_f3<-tableGrob(trans_t2p3, theme=mytheme,rows=NULL)
 ####flextable
gt3_2=autofit(flextable(trans_t2p3)%>%set_caption(caption = "Table 2C: CCS/HiFi Stats. This table contains metrics for all HiFi (CCS) reads generated by the run.")%>% theme_zebra()%>%bg(bg = "#44B29C", part = "header"))
 }
 }
#Display appropriate tables
if(exists('gt1') & exists('gt2') & exists('gt3_2')){
  flextable_to_rmd(gt1)
  flextable_to_rmd(gt2)
  flextable_to_rmd(gt3_2)
  write.csv(trans_t2p1, file=paste0(csvout, "/table2a.csv"))
  write.csv(trans_t2p2, file=paste0(csvout, "/table2b.csv"))
  write.csv(trans_t2p3, file=paste0(csvout, "/table2c.csv"))
} else if (exists('gt1') & exists('gt2') & exists('gt3_2')==FALSE){
  flextable_to_rmd(gt1)
  flextable_to_rmd(gt2)
  write.csv(trans_t2p1, file=paste0(csvout, "/table2a.csv"))
  write.csv(trans_t2p2, file=paste0(csvout, "/table2b.csv"))
} else if (exists('gt1') & exists('gt3_2') & exists('gt2')==FALSE){
  flextable_to_rmd(gt1)
  flextable_to_rmd(gt3_2)
  write.csv(trans_t2p1, file=paste0(csvout, "/table2a.csv"))
  write.csv(trans_t2p3, file=paste0(csvout, "/table2c.csv"))
} else if ( exists('gt2') & exists('gt3_2') & exists('gt1')==FALSE){
  flextable_to_rmd(gt2)
  flextable_to_rmd(gt3_2)
  write.csv(trans_t2p2, file=paste0(csvout, "/table2b.csv"))
  write.csv(trans_t2p3, file=paste0(csvout, "/table2c.csv"))
}

```

```{r Table3, echo=FALSE,fig.width=14, fig.height=14, message=FALSE, fig.align='center', warning=FALSE, results='asis'}
##Reset grid variables
rm(gt1)
#rm(gt2)
if(file.exists(demuxpath)){
text3=readLines(demuxpath)
chunksplit <- which(!nzchar(text3))[1]
if(is.na(chunksplit)){
  chunksplit=length(text3)+1
}
chunk1=text3[1:chunksplit -1 ]
if(length(chunk1)>5){
i1 <- grep("Demux Stats:", chunk1)+1
cellspec=text3[grepl(dropid,text3)==T]
if(length(cellspec)!=0){
cnames=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk1[i1])), header=T, sep="|", row.names = NULL, check.names=FALSE)
    i2 <-grep("total_demux_bases", cellspec)-1
     if(is.na(chunksplit)==F){
       totals_line=i2+1
    t3p1_ll=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", cellspec[totals_line])), header=F, sep="|", row.names = NULL, check.names=FALSE)
    t3_p1=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", cellspec[1:i2])), header=F, sep="|", row.names = NULL, check.names=FALSE)
    colnames(t3_p1)=colnames(cnames)
    t3_p1$X=NULL
    t3_p1$X.1=NULL
    t3_filt1t=t3_p1[t3_p1$cell==dropid,]
    #Deal with blank columns
    empty_columns <- colSums(is.na(t3_filt1t) | t3_filt1t == "") == nrow(t3_filt1t)
    t3_filt1=t3_filt1t[, !empty_columns]
    ####Add last line
    lline=data.frame("Totals", "-", "-","-",t3p1_ll[6],t3p1_ll[10], t3p1_ll[4],t3p1_ll[8], "-","-","-","-","-","-","-","-","-")
    cnames=colnames(t3_filt1)
    colnames(lline)=cnames
    t3_filt1ll=rbind(t3_filt1,lline)
    gt3=autofit(flextable(t3_filt1ll)%>%set_caption(caption = "Table 3A: SubRead Demultiplexing Stats. This table lists subread metrics for all samples demultiplexed as part of the run.")%>% theme_zebra()%>%bg(bg = "#44B29C", part = "header"))%>%fit_to_width(max_width = 8.5)
    flextable_to_rmd(gt3)
    write.csv(t3_filt1ll, file=paste0(csvout, "/table3a.csv"))
}}}
i1 <- grep("CCS Demux Stats:", text3)+1
i2 <-max(grep("total_demux_ccs_bp",text3))
#i2 <-grep("total_demux_ccs_bp",text3)[1]
if(identical(i1, numeric(0))!=T){
cellspec<-text3[i1:i2]
chunk2<-text3[i1]
t3f2_ll=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", cellspec[length(cellspec)])), header=F, sep="|", row.names = NULL,check.names=FALSE)
    lline=data.frame("Totals", "-", "-","-",t3f2_ll[6],t3f2_ll[10], t3f2_ll[4],t3f2_ll[8], "-","-","-","-","-","-","-","-","-")
    #CSS Demux Stats
    cnames2=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk2)), header=T, sep="|", row.names = NULL,check.names=FALSE)
    #cellspec2=cellspec[1:length(cellspec)-1]
    cellspec2=cellspec[grep(dropid, cellspec)]
    t3_p2=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", cellspec2[1:length(cellspec2)-1])), header=F, sep="|", row.names = NULL, col.names = colnames(cnames2), check.names=FALSE)
    t3_p2$X=NULL
    t3_p2$X.1=NULL
    t3_filt2t=t3_p2[t3_p2$cell==dropid,]
    if(nrow(t3_filt2t)>0){
    empty_columns <- colSums(is.na(t3_filt2t) | t3_filt2t == "") == nrow(t3_filt2t)
    t3_filt2=t3_filt2t[, !empty_columns]
    cnames=colnames(t3_filt2)
    colnames(lline)=cnames
    #Remove NA colnames
    lline<-lline[!is.na(names(lline))]
    t3_filt2ll=rbind(t3_filt2,lline)
    #flextable
    gt3=autofit(flextable(t3_filt2ll)%>%set_caption(caption = "Table 3B: HiFi (CCS) Demultiplexing Stats. This table lists HiFi (CCS) metrics for all samples demultiplexed as part of the run.")%>% theme_zebra()%>%bg(bg = "#44B29C", part = "header"))%>%fit_to_width(max_width = 8.5)
    flextable_to_rmd(gt3)
    write.csv(t3_filt2ll, file=paste0(csvout, "/table3b.csv"))
}}}
```


```{r Table3Fig, echo=FALSE,fig.width=10, fig.height=12, message=FALSE, fig.align='center', warning=FALSE, fig.cap="Figures show SubRead Demultiplexing Stats for bases (top) and reads (bottom). "}
if(file.exists(demuxpath)){
require(scales)
#Barplots for Demux Stats of they exist
if(exists('t3_filt1')){
t3_filt1$bases= as.numeric(gsub(",", "", t3_filt1$bases))
t3_filt1$reads= as.numeric(gsub(",", "", t3_filt1$reads))
tab3f1=ggplot(data=t3_filt1, aes(x=`bases`, y=sample)) +geom_bar(stat="identity", color="#44B29C", fill="#44B29C")+ coord_flip()+ylab("Samples")+ xlab("bases")+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+ scale_x_continuous(labels = comma)
tab3f2=ggplot(data=t3_filt1, aes(x=`reads`, y=sample)) +geom_bar(stat="identity", color="#334E5C", fill="#334E5C")+ coord_flip()+ylab("Samples")+ xlab("reads")+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+ scale_x_continuous(labels = comma)
grid.arrange(tab3f1,tab3f2)
}}
```

```{r Table3Figb, echo=FALSE,fig.width=10, fig.height=12, message=FALSE, fig.align='center', warning=FALSE, fig.cap="Figures show HiFi (CCS) Demultiplexing Stats for bases (top) and reads (bottom)."}
#Barplots for CCS Demux Stats of they exist
if(file.exists(demuxpath)){
  if(exists('t3_filt2')){
    t3_filt2$bases= as.numeric(gsub(",", "", t3_filt2$bases))
t3_filt2$reads= as.numeric(gsub(",", "", t3_filt2$reads))

    tab3f3=ggplot(data=t3_filt2, aes(x=`reads`, y=sample)) +geom_bar(stat="identity",color="#FFCD00", fill="#FFCD00")+ coord_flip()+ylab("Samples")+ xlab("reads")+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+ scale_x_continuous(labels = comma)

tab3f4=ggplot(data=t3_filt2, aes(x=`bases`, y=sample)) +geom_bar(stat="identity",color="#C8102E", fill="#C8102E")+ coord_flip()+ylab("Samples")+ xlab("bases")+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+ scale_x_continuous(labels = comma)

grid.arrange(tab3f4,tab3f3)

}}

```

```{r Table4, echo=F,fig.width=14, fig.height=12, message=F, fig.align='center',  fig.asp=1, warning=FALSE, results='asis'}
#Parse json
mytheme1 <- gridExtra::ttheme_default(base_size=16, core = list(fg_params=list(cex = 0.75, fontface="plain")),colhead = list(fg_params=list(cex = 0.75, fontface="bold")),rowhead = list(fg_params=list(cex = 0.75, fontface="bold")))
holdirs=list.dirs(path = paste0(basep,runid), full.names = TRUE, recursive = F)
cellp=holdirs[grepl(dropid, holdirs)==T]
cfjsonp=paste0(cellp,"/Filter/dataset-reports/outputs/control.report.json")
winners <- fromJSON(cfjsonp, flatten=TRUE)
jtab=winners$attributes
jtab$value=format(round(as.numeric(jtab$value), 2),scientific=FALSE,big.mark=",")
jtab$id<-NULL
names(jtab)<-c("Parameter", "Value")

gt1=autofit(flextable(jtab)%>%set_caption(caption = "Table 4: Table contains metrics for spike-in controls on the run.")%>%theme_zebra()%>%bg(bg = "#44B29C", part = "header"))
flextable_to_rmd(gt1)

write.csv(jtab, file=paste0(csvout, "/table4.csv"))
```

```{r Figures1, echo=FALSE,fig.align='center', out.width='0.75\\linewidth', message=FALSE}
holdirs=list.dirs(path = paste0(basep,runid), full.names = TRUE, recursive = F)
cellp=holdirs[grepl(dropid, holdirs)==T]
figpathp="/Filter/dataset-reports/cromwell-job/"
hp1=paste0(cellp,figpathp)
fp2=list.dirs(hp1, full.names=T,recursive=F)
figpath=paste0(hp1,"/call-import_dataset_reports/execution/")

fig1n=paste0(figpath,"raw_read_length_plot.png")
fig2n=paste0(figpath,"subread_lengths.png")
fig3n=paste0(figpath,"hexbin_length_plot.png")
fig4n=paste0(figpath,"base_yield_plot.png")

```

```{r Figures1a, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="High-Quality Region Filtering Evaluation - This plot compares read count and read length for pre- and post-filtered reads."}
if(file.exists(fig1n)){
  ggdraw() + draw_image(fig1n)
}
```

```{r Figures1b, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="SubRead Length Distribution"}
if(file.exists(fig2n)){
ggdraw() + draw_image(fig2n)
}
```

```{r Figures1c, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="Longest Subread vs Polymerase Read Length - This figure plots, for each ZMW (well) on the SMRT Cell, the longest subread length vs the polymerase length. Density (count) is color coded."}
if(file.exists(fig3n)){
  ggdraw() + draw_image(fig3n)
}
```

```{r Figures1d, echo=FALSE,fig.align='center', out.width='100%', message=FALSE, fig.cap="Base Yield Density"}
if(file.exists(fig4n)){
  ggdraw() + draw_image(fig4n)
}
```

```{r Figures2, echo=FALSE,fig.align='center', out.width='0.75\\linewidth', message=FALSE, fig.cap="This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents."}
jsonfig1="readlength_plot.png"
jsonfig2="concordance_plot.png"

fig5n=paste0(figpath,jsonfig1)
fig6n=paste0(figpath,jsonfig2)
```

```{r Figures1e, echo=FALSE,fig.align='center', fig.width=14, fig.height=12,message=FALSE, fig.cap="Read Length Plot (left) and concordance plot (right)", warning=FALSE}
if(file.exists(fig5n)==T & file.exists(fig6n)==T){
p1 <- ggdraw() + draw_image(fig5n)
p2 <- ggdraw() + draw_image(fig6n)
gg=plot_grid(p1, p2)
}
gg
```

```{r rawqc, echo=FALSE, message=FALSE, warning=FALSE, fig.align="center", results='asis'}
#runinfo sometimes has 2 lines before relevant text. Normally just one, old runs might have 2.

Str1="Raw Data QC - Contamination Screen Results"

Str2="For each library, a randomly selected subset of reads are screened for contamination by MegaBLAST against a local copy of the NCBI nucleotide (nt) database. The purpose of this screen is only to identify potential contamination in the sample, not to fully characterize the resulting data set."
postpend="/Demux/QC/"
dqcp=paste0(cellp,postpend)
Flag=FALSE
if(dir.exists(dqcp)){
#cat("\\pagebreak")
ldqcp=list.files(path=dqcp,pattern="topHits.txt", full.names=T)
if(length(ldqcp)!=0){
for(f in ldqcp){
  if(file.size(f) != 0L){
    text4=readLines(f)
    #Adding genus:
    genus=sub("^.+:", "", text4[1])
    species=gsub("\t","",sub("^.+:", "", text4[2]))
    stringpp<-paste0("The expected genus is ", genus, "and the expected species is ",species,".")
    Flag=TRUE
  }}
    cat(paste0("# ",Str1,"\n"))
    cat(paste0("### ",Str2))
}}

postpend="/QC/"
dqcp=paste0(cellp,postpend)
if(dir.exists(dqcp)){
ldqcp=list.files(path=dqcp,pattern="topHits.txt", full.names=T)
#cat("\\pagebreak")
if(length(ldqcp)!=0){
for(f in ldqcp){
  if(file.size(f) != 0L & Flag==FALSE){
    text4=readLines(f)
    #Adding genus:
    genus=sub("^.+:", "", text4[1])
    species=gsub("\t","",sub("^.+:", "", text4[2]))
    stringpp<-paste0("The expected genus is ", genus, " and the expected sprecies is ",species,".")
   }}
     cat(paste0("# ",Str1,"\n"))
    cat(paste0("## ",Str2))
    #cat(stringpp) 
  }}
```

```{r Table5, echo=F,fig.width=12, fig.height=12, message=F, fig.align='center', results='hide', fig.pos='H', warning=FALSE, warning=FALSE}
##Check for demux QC
postpend="/Demux/QC/"
dqcp=paste0(cellp,postpend)
if(dir.exists(dqcp)){
ldqcp=list.files(path=dqcp,pattern="topHits.txt", full.names=T)
if(length(ldqcp)!=0){
for(f in ldqcp){
  if(file.size(f) != 0L){
    text4=readLines(f)
    #Adding genus:
    genus=sub("^.+:", "", text4[1])
    species=gsub("\t","",sub("^.+:", "", text4[2]))
    chunksplit <- which(!nzchar(text4))
    start1=chunksplit[1]+1
    end1=chunksplit[2]-1
    chunk1=text4[start1:end1]
    t4p1=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk1)), header=F, sep="\t", row.names = NULL, check.names=FALSE)
    samplename=basename(f)
    sname=gsub( ".medianSubset.fasta.blast.topHits.txt","", samplename)
    ###make table grob
    names(t4p1)<-NULL
    t4_f1<-tableGrob(t4p1, theme=mytheme,rows=NULL)
    table4title1=strwrap(paste0("Table 5: Summary Results for ",sname,". This table lists metrics for the expected genus & species, and the top other genus and species."),width = 175, simplify = F)
    title_wrapped<-sapply(table4title1, paste, collapse = "\n")
    h1 <- grobHeight(t4_f1)
    w1 <- grobWidth(t4_f1)
    title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust=0)
    gt1 <- gTree(children=gList(t4_f1,title1))
    write.csv(t4p1, file=paste0(csvout, "/table5a.csv"))
    ###flextable implementation
    #colnames(t4p1)<-c("Parameter","Value")
    #gt1=autofit(flextable(t4p1)%>%theme_zebra()%>%bg(bg = "#44B29C", part = "header"))%>% as_raster()
    #flextable_to_rmd(gt1)
    start2=end1+2
    end2=chunksplit[3]-1
    chunk2=text4[start2:end2]
    t4p2=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk2)), header=F, sep="\t", row.names = NULL, check.names=FALSE)
    names(t4p2)<-NULL
    ###make table grob
    t4_f2<-tableGrob(t4p2, theme=mytheme,rows=NULL)
    table4title1=strwrap(paste0("Table 5: Summary Results for ",sname,". This table lists metrics for the expected genus ",genus," & species ",species,", and the top other genus and species."),width = 165, simplify = F)
    title_wrapped<-sapply(table4title1, paste, collapse = "\n")
    h1 <- grobHeight(t4_f2)
    w1 <- grobWidth(t4_f2)
    title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust=0)
    gtt1<-arrangeGrob(t4_f1,t4_f2, ncol=2)
    gt2 <- gTree(children=gList(gtt1,title1))
    write.csv(t4p2, file=paste0(csvout, "/table5a.csv"))
    ###flextable implementation
    #colnames(t4p2)<-c("Parameter","Value")
    #gt2=autofit(flextable(t4p2)%>%theme_zebra()%>%bg(bg = "#44B29C", part = "header"))%>% as_raster()
    #flextable_to_rmd(gt1)
    ###Arrange flextable
    #gg1=ggplot() + theme_void() + annotation_custom(rasterGrob(gt1), xmin=-Inf, xmax=Inf, ymin=-Inf, ymax=Inf)
    #gg2=ggplot() + theme_void() + annotation_custom(rasterGrob(gt2), xmin=-Inf, xmax=Inf, ymin=-Inf, ymax=Inf)
    #cowplot::plot_grid(gg1, gg2, nrow = 1, ncol = 2, rel_heights = c(3, 1) )
    start3=chunksplit[3]+3
    chunk3=text4[start3:length(text4)]
    t4p3=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk3)), header=T, sep="\t", row.names = NULL, check.names=FALSE)
    if(nrow(t4p3)>1){
    # ###make table grob
    # t4_f3<-tableGrob(t4p3[1:25,], theme=mytheme,rows=NULL)
    # table4title1=strwrap(paste0("Table 6: Contamination Screen Summary for ",sname,". This table lists the top 25 matches for sample the library sequenced, ranked by the number of reads for each hit."),width = 175, simplify = F)
    # title_wrapped<-sapply(table4title1, paste, collapse = "\n")
    # h1 <- grobHeight(t4_f3)
    # w1 <- grobWidth(t4_f3)
    # title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust=0)
    # gt1 <- gTree(children=gList(t4_f3,title1))
    ###flextable
    gt1=autofit(flextable(t4p3[1:25,])%>%set_caption(caption = paste0("Table 6: Contamination Screen Summary for ",sname,". This table lists the top 25 matches for sample the library sequenced, ranked by the number of reads for each hit."))%>%theme_zebra()%>%bg(bg = "#44B29C", part = "header"))%>% as_raster()
    #flextable_to_rmd(gt1)
    gg1=ggplot() + theme_void() + annotation_custom(rasterGrob(gt1), xmin=-Inf, xmax=Inf, ymin=-Inf, ymax=Inf)
    grid.arrange(arrangeGrob(gt2), 
             arrangeGrob(gg1), 
             nrow=2)
    write.csv(t4p3, file=paste0(csvout, "/table6.csv"))
    
    } else {
      grid.arrange(gt2)
    }
}}}
}
##Check for non demux QC
postpend="/QC/"
dqcp=paste0(cellp,postpend)
if(dir.exists(dqcp)){
ldqcp=list.files(path=dqcp,pattern="topHits.txt", full.names=T)
if(length(ldqcp)!=0){
for(f in ldqcp){
  #add condition to check if f is empty
  if(file.size(f) != 0L){
    text4=readLines(f)
    chunksplit <- which(!nzchar(text4))
    start1=chunksplit[1]+1
    end1=chunksplit[2]-1
    chunk1=text4[start1:end1]
    t4p1=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk1)), header=F, sep="\t", row.names = NULL, check.names=FALSE)
    samplename=basename(f)
    sname=gsub( ".medianSubset.fasta.blast.topHits.txt","", samplename)
    sname=gsub("\\..*","",sname)
    ###make table grob
    names(t4p1)<-NULL
    t4_f1<-tableGrob(t4p1, theme=mytheme,rows=NULL)
    table4title1=strwrap(paste0("Table 5: Summary Results for ",sname,". This table lists metrics for the expected genus ",genus," & species ",species,", and the top other genus and species."),width = 175, simplify = F)
    title_wrapped<-sapply(table4title1, paste, collapse = "\n")
    h1 <- grobHeight(t4_f1)
    w1 <- grobWidth(t4_f1)
    title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust=0)
    gt1 <- gTree(children=gList(t4_f1,title1))
    write.csv(t4p1, file=paste0(csvout, "/table5a.csv"))
    start2=end1+2
    end2=chunksplit[3]-1
    chunk2=text4[start2:end2]
    t4p2=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk2)), header=F, sep="\t", row.names = NULL, check.names=FALSE)
    names(t4p2)<-NULL
    ###make table grob
    t4_f2<-tableGrob(t4p2, theme=mytheme,rows=NULL)
    table4title1=strwrap(paste0("Table 5: Summary Results for ",sname,". This table lists metrics for the expected genus ",genus," & species ",species,", and the top other genus and species."),width = 165, simplify = F)
    title_wrapped<-sapply(table4title1, paste, collapse = "\n")
    h1 <- grobHeight(t4_f2)
    w1 <- grobWidth(t4_f2)
    title1 <- textGrob(title_wrapped, gp=gpar(fontsize=12), y=unit(0.6,"npc")+ 0.75*h1, vjust=0)
    gtt1=arrangeGrob(t4_f1,t4_f2, ncol=2)
    gt2 <- gTree(children=gList(gtt1,title1))
    write.csv(t4p2, file=paste0(csvout, "/table5b.csv"))
    start3=chunksplit[3]+3
    chunk3=text4[start3:length(text4)]
    t4p3=read.csv(text=gsub("\\.,", ",", sub("\\s*,$", "", chunk3)), header=T, sep="\t", row.names = NULL, check.names=FALSE)
    if(nrow(t4p3)>1){
    ###flextable
    gt1=autofit(flextable(t4p3[1:25,])%>%set_caption(caption = paste0("Table 6: Contamination Screen Summary for ",sname,". This table lists the top 25 matches for sample the library sequenced, ranked by the number of reads for each hit."))%>%theme_zebra()%>%bg(bg = "#44B29C", part = "header"))%>% as_raster()
    #flextable_to_rmd(gt1)
    gg1=ggplot() + theme_void() + annotation_custom(rasterGrob(gt1), xmin=-Inf, xmax=Inf, ymin=-Inf, ymax=Inf)
    grid.arrange(arrangeGrob(gt2), arrangeGrob(gg1), nrow=2)
    write.csv(t4p3, file=paste0(csvout, "/table6.csv"))
    } else {
      grid.arrange(gt2)
    }

}}}}

```
